core/shared/src/main/scala/org/http4s/multipart/Part.scala

```diff
@@ -31,11 +31,62 @@ import org.typelevel.ci._
 import java.io.File
 import java.io.InputStream
 import java.net.URL
+import java.nio.charset.StandardCharsets
+import java.nio.charset.{Charset => NioCharset}
 
 final case class Part[F[_]](headers: Headers, body: Stream[F, Byte]) extends Media[F] {
-  def name: Option[String] = headers.get[`Content-Disposition`].flatMap(_.parameters.get(ci"name"))
+
+  /** This part's name from its Content-Disposition header, decoded as
+    * UTF-8.
+    */
+  def name: Option[String] =
+    name(StandardCharsets.UTF_8)
+
+  /** This part's name from its Content-Disposition header, decoded as
+    * the specified charset.
+    */
+  def name(charset: NioCharset): Option[String] =
+    contentDispositionParam(ci"name", charset)
+
+  /** This part's name from its Content-Disposition header, decoded from
+    * raw bytes with the specified function.
+    */
+  def nameDecoded(f: Array[Byte] => Option[String]): Option[String] =
+    contentDispositionParamDecoded(ci"name", f)
+
+  /** This part's filename from its Content-Disposition header, decoded
+    * as UTF-8.
+    */
   def filename: Option[String] =
-    headers.get[`Content-Disposition`].flatMap(_.parameters.get(ci"filename"))
+    filename(StandardCharsets.UTF_8)
+
+  /** This part's filename from its Content-Disposition header, decoded
+    * as the specified charset.
+    */
+  def filename(charset: NioCharset): Option[String] =
+    contentDispositionParam(ci"filename", charset)
+
+  /** This part's filename from its Content-Disposition header, decoded
+    * from raw bytes with the specified function.
+    */
+  def filenameDecoded(f: Array[Byte] => Option[String]): Option[String] =
+    contentDispositionParamDecoded(ci"filename", f)
+
+  private def contentDispositionParam(name: CIString, charset: NioCharset): Option[String] =
+    charset match {
+      case StandardCharsets.ISO_8859_1 =>
+        headers.get[`Content-Disposition`].flatMap(_.parameters.get(name))
+      case charset =>
+        contentDispositionParamDecoded(name, bytes => Some(new String(bytes, charset)))
+    }
+
+  private def contentDispositionParamDecoded(
+      name: CIString,
+      f: Array[Byte] => Option[String],
```

Comment:
This feels like `f: Array[Byte] => Either[DecodeFailure, String]`. I don't have any examples on hand to enhance this take, though

Reply:
The main reason to use this without a `Charset` would be percent encoding, which can fail.

The charset encoders will render a character to represent improper encodings rather than fail.  We could make this a pure function and make it consistent with that.  I'd rather make an `Either` and show that it failed, but I don't necessarily know what to do with the Either when there's already an Option wrapper.  Either[^1] way, maybe an Option that gets silently flattened is the worst thing we can do.

[^1]: no pun intended